lib/Runtime/Library/Utf8String.h

```diff
@@ -0,0 +1,142 @@
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#pragma once
+
+#include "Codex/Utf8Codex.h"
+
+namespace Js
+{
+    class Utf8String : public JavascriptString
+    {
+    private:
+        typedef struct {
+            FieldNoBarrier(size_t) length;
+            Field(char*) buffer;
+        } PrefixedUtf8String;
+        Field(PrefixedUtf8String*) utf8String;
+
+    protected:
+        DEFINE_VTABLE_CTOR(Utf8String, JavascriptString);
+
+    private:
+        
+        void SetUtf8Buffer(_In_reads_(utf8Length) char* buffer, size_t utf8Length)
+        {
+            this->utf8String = RecyclerNew(this->GetRecycler(), PrefixedUtf8String);
+            this->utf8String->length = utf8Length;
+            this->utf8String->buffer = buffer;
+        }
+
+    public:
+
+        Utf8String(_In_ JavascriptString* originalString, _In_reads_(utf8Length) char* buffer, size_t utf8Length, _In_ StaticType* type) :
+            JavascriptString(type),
+            utf8String(nullptr)
+        {
+            SetUtf8Buffer(buffer, utf8Length);
+
+            this->SetLength(originalString->GetLength());
+            this->SetBuffer(originalString->UnsafeGetBuffer());
+        }
+
+        Utf8String(_In_reads_(utf8Length) char* buffer, size_t utf8Length, _In_ StaticType* type) :
+            JavascriptString(type),
+            utf8String(nullptr)
+        {
+            SetUtf8Buffer(buffer, utf8Length);
+
+            charcount_t utf16Length = 0;
+            utf8::DecodeOptions opts = utf8::DecodeOptions::doDefault;
+            LPCUTF8 buf = reinterpret_cast<LPCUTF8>(buffer);
+            LPCUTF8 end = buf + utf8Length;
+            while (buf < end)
+            {
+                if ((*buf & 0x80) == 0)
+                {
+                    // Single byte character
+                    utf16Length++;
+                    buf++;
+                }
+                else
+                {
+                    // Decode a single utf16 character, and increment the pointer
+                    // to the start of the next character.
+                    char16 c1 = *buf;
+                    ++buf;
+                    utf8::DecodeTail(c1, buf, end, opts);
+                    utf16Length++;
+                }
+            }
+
+            this->SetLength(utf16Length);
+            this->SetBuffer(nullptr);
+        }
+
+
+        size_t Utf8Length() const
+        {
+            return this->utf8String->length;
+        }
+
+        const char* Utf8Buffer() const
+        {
+            return this->utf8String->buffer;
+        }
+
+        const char16* GetSz() override
+        {
+            if (this->IsFinalized())
+            {
+                return this->UnsafeGetBuffer();
+            }
+
+            // TODO: This is currently wrong in the presence of unmatched surrogate pairs.
```

Comment:
Nit: it might be nice to move this comment down to right before DecodeUnitsIntoAndNullTerminateNoAdvance, because in this position I originally though it was talking about the allocation size being wrong (which would be Very Bad).

Reply:
I think that's also not the 'correct' place for this comment; been a while since I wrote it. The actual issue is with the encoding into utf8, and determining how to decode it. As things stand I'm using 'actual' utf8 which doesn't allow lone surrogate halves, but we could instead use 'cesu-8' which does allow that.

Reply:
As far as I know canonical utf-8 doesn’t encode surrogate halves at all (paired or otherwise)—the character is encoded directly.  If you need to preserve the surrogates independently I think cesu-8 is your only option.